File: android/src/main/java/expo/modules/connect4solver/Connect4SolverModule.kt
package expo.modules.connect4solver

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

import expo.modules.kotlin.records.Record
import expo.modules.kotlin.records.Field

class Connect4SolverModule : Module() {
    private external fun nativeCreateCache(width: Int, height: Int, sizeBytes: Double, align: Int, wrap: Boolean): String
    private external fun nativeDestroyCache(cachePtrStr: String)
    private external fun nativeCreateSolver(width: Int, height: Int, cachePtrStr: String, align: Int, wrap: Boolean): String
    private external fun nativeDestroySolver(solverPtrStr: String, width: Int, height: Int, align: Int, wrap: Boolean)
    private external fun nativeStop(solverPtrStr: String, width: Int, height: Int, align: Int, wrap: Boolean)
    private external fun nativeCreateBookFromBuffer(width: Int, height: Int, base64: ByteArray): String
    private external fun nativeDestroyBook(width: Int, height: Int, bookPtrStr: String)
    private external fun nativeAnalyze(solverPtrStr: String, position: String, threads: Int, width: Int, height: Int, bookPtrStr: String, align: Int, wrap: Boolean): IntArray?
    private external fun nativeSolve(solverPtrStr: String, position: String, threads: Int, width: Int, height: Int, bookPtrStr: String, align: Int, wrap: Boolean): IntArray?

    class AnalyzeArgs : Record {
        @Field var solverPtr: String = ""
        @Field var pos: String = ""
        @Field var threads: Int = 1
        @Field var w: Int = 7
        @Field var h: Int = 6
        @Field var bookPtr: String = ""
        @Field var align: Int = 4
        @Field var wrap: Boolean = false
    }

    companion object {
        init {
            System.loadLibrary("connect4solver")
        }
    }

    override fun definition() = ModuleDefinition {
        Name("Connect4Solver")

        Function("createCache") { width: Int, height: Int, sizeBytes: Double, align: Int, wrap: Boolean ->
            return@Function nativeCreateCache(width, height, sizeBytes, align, wrap)
        }

        Function("destroyCache") { cachePtrStr: String ->
            nativeDestroyCache(cachePtrStr)
        }

        Function("createSolver") { width: Int, height: Int, cachePtrStr: String, align: Int, wrap: Boolean ->
            return@Function nativeCreateSolver(width, height, cachePtrStr, align, wrap)
        }

        Function("destroySolver") { solverPtrStr: String, width: Int, height: Int, align: Int, wrap: Boolean ->
            nativeDestroySolver(solverPtrStr, width, height, align, wrap)
        }

        Function("stop") { solverPtrStr: String, width: Int, height: Int, align: Int, wrap: Boolean ->
            nativeStop(solverPtrStr, width, height, align, wrap)
        }

        Function("createBookFromBuffer") { width: Int, height: Int, base64Str: String ->
            val decodedBytes = android.util.Base64.decode(base64Str, android.util.Base64.DEFAULT)
            return@Function nativeCreateBookFromBuffer(width, height, decodedBytes)
        }

        Function("destroyBook") { width: Int, height: Int, bookPtrStr: String ->
            nativeDestroyBook(width, height, bookPtrStr)
        }

        AsyncFunction("analyze") { args: AnalyzeArgs ->
            val result = nativeAnalyze(args.solverPtr, args.pos, args.threads, args.w, args.h, args.bookPtr, args.align, args.wrap)
            if (result != null) {
                return@AsyncFunction result.toList()
            } else {
                throw Exception("Unsupported board size")
            }
        }

        AsyncFunction("solve") { args: AnalyzeArgs ->
            val result = nativeSolve(args.solverPtr, args.pos, args.threads, args.w, args.h, args.bookPtr, args.align, args.wrap)
            if (result != null) {
                return@AsyncFunction result.toList()
            } else {
                throw Exception("Unsupported board size")
            }
        }

    }
}

File: ios/Connect4SolverModule.swift
import ExpoModulesCore

struct AnalyzeArgs: Record {
    @Field var solverPtr: String
    @Field var pos: String
    @Field var threads: Int
    @Field var w: Int
    @Field var h: Int
    @Field var bookPtr: String
    @Field var align: Int
    @Field var wrap: Bool
}

public class Connect4SolverModule: Module {
    public func definition() -> ModuleDefinition {
        Name("Connect4Solver")

        Function("createCache") { (width: Int, height: Int, sizeBytes: Double, align: Int, wrap: Bool) -> String in
            return Connect4SolverWrapper.createCache(Int32(width), height: Int32(height), sizeBytes: sizeBytes, align: Int32(align), wrap: wrap)
        }

        Function("destroyCache") { (cachePtrStr: String) in
            Connect4SolverWrapper.destroyCache(cachePtrStr)
        }

        Function("createSolver") { (width: Int, height: Int, cachePtrStr: String, align: Int, wrap: Bool) -> String in
            return Connect4SolverWrapper.createSolver(Int32(width), height: Int32(height), cachePtrStr: cachePtrStr, align: Int32(align), wrap: wrap)
        }

        Function("destroySolver") { (solverPtrStr: String, width: Int, height: Int, align: Int, wrap: Bool) in
            Connect4SolverWrapper.destroySolver(solverPtrStr, width: Int32(width), height: Int32(height), align: Int32(align), wrap: wrap)
        }

        Function("stop") { (solverPtrStr: String, width: Int, height: Int, align: Int, wrap: Bool) in
            Connect4SolverWrapper.stop(solverPtrStr, width: Int32(width), height: Int32(height), align: Int32(align), wrap: wrap)
        }

        Function("createBookFromBuffer") { (width: Int, height: Int, base64Str: String) -> String in
            return Connect4SolverWrapper.createBook(fromBuffer: Int32(width), height: Int32(height), base64Str: base64Str)
        }

        Function("destroyBook") { (width: Int, height: Int, bookPtrStr: String) in
            Connect4SolverWrapper.destroyBook(Int32(width), height: Int32(height), bookPtrStr: bookPtrStr)
        }

        AsyncFunction("analyze") { (args: AnalyzeArgs, promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                if let result = Connect4SolverWrapper.analyze(args.solverPtr, position: args.pos, threads: Int32(args.threads), width: Int32(args.w), height: Int32(args.h), bookPtrStr: args.bookPtr, align: Int32(args.align), wrap: args.wrap) {
                    promise.resolve(result)
                } else {
                    promise.reject("UNSUPPORTED_SIZE", "Unsupported board size")
                }
            }
        }

        AsyncFunction("solve") { (args: AnalyzeArgs, promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                if let result = Connect4SolverWrapper.solve(args.solverPtr, position: args.pos, threads: Int32(args.threads), width: Int32(args.w), height: Int32(args.h), bookPtrStr: args.bookPtr, align: Int32(args.align), wrap: args.wrap) {
                    promise.resolve(result)
                } else {
                    promise.reject("UNSUPPORTED_SIZE", "Unsupported board size")
                }
            }
        }

    }
}
